### ISSUE TITLE
Having issues when saving objects in an array of an Astronomy.Class

### ISSUE THREAD

**Issue by banglashi** [jagi/meteor-astronomy#6]
Hi @jagi 

when trying to save an array of objects, it works the first time, second time not.

You can use the same repo here: https://github.com/banglashi/astronomy-serverside-transform

Than run following on the console:

var newChat = new Chat()
undefined
newChat.title = "test";
"test"
newChat.members.push({title:'test'})
1
newChat.save()
"pz5FLFERiXgFfjkzJ"
newChat.members.push({title:'test2'})
2
newChat.save()
false

Any ideas? Is this perhaps still in development? Thanks and regards.

**Comment by jagi** [comment]
Yes I know about that issue. The problem is that when you modify array Astronomy object doesn't know that something has changed. It can only detect changes for primitive objects because I use `===` comparison. In that case `array` and `modified array` are the same objects in memory (it's the same reference). I have solution for that and I will implement it soon. However there is a workaround, ugly but work until I implement correct behavior.

```js
chat.members = chat.members.slice();
chat.members.push({title: 'title'});
chat.save();
```

**Comment by jagi** [comment]
It's more complex problem than I thought. I will post few possible solutions.

**SOLUTION 1**

```js
chat.members = chat.member;
// Line above calls "set" method which simplifying looks like follows
function(value) {
  return value.slice(); // Clone object instead assigning reference
}
// After that "chat" object will already know that "members" field is modified so we can do whatever we want to this field like for example.
chat.members.push({title: 'title'});
// Calling...
chat.save();
// ... will update the "members" field in the collection.
```

It's not the best solution but the shortest one. Disadvantages?
- We have to remember about writing `chat.members = chat.members` to copy value
- We can't assign reference if we would like to. Let's say we have two objects `chat1` and `chat2` and we want to make assignment by reference `chat1.members = chat2.members` - it will not work. For such situations we would need to create some special field type like `reference`. But I don't think it's good idea.

**SOLUTION 2**

Right now every Astronomy document has two internal objects `_values` and `_modified`. When we change some value it's stored in the `_modified` object. When we call the `save()` method it takes modified fields (using `getModified()` method) and saves only those fields. It's simple comparison (`===`) of the `_modified` and `_values` objects. That's it.

In the new solution, I would create two internal objects named `_original` and `_values` in every document. The difference is that both object would store the same data at the beginning. However `_original` object would be list of copies not references of all fields in the `_values` object. Every modification would take place on the `_values` object. And now is the best part. On the `save()` method call it would also run `getModified()` method but it would check what fields differ using `EJSON.equals()` method. Thanks to that we wouldn't have to care if we are working on the reference or original object. It would always compare copies on save.

And that's is the solution I will implement in following days.

**Comment by banglashi** [comment]
sounds perfect to me. Thanks for looking into it.

**Comment by jagi** [comment]
Today or tomorrow feature will be ready, so stay tuned.

**Comment by jagi** [comment]
@banglashi  I've made a fix in the version 0.5.1. It's already published and you can check it out. I would be glad if you could test it out and let me know if everything works as expected. Thanks.
